subsys/bluetooth/audio/bass.c

```diff
@@ -0,0 +1,1373 @@
+	    recv_state->pa_sync_state == BT_BASS_PA_STATE_INFO_REQ) {
+		return;
+	}
+
+	if (conn != NULL && pa_past &&
+	    BT_FEAT_LE_PAST_SEND(conn->le.features) &&
+	    BT_FEAT_LE_PAST_RECV(bt_dev.le.features)) {
+		bass_pa_sync_past(conn, state);
+	} else {
+		bass_pa_sync_no_past(state);
+	}
+}
+
+static int bass_add_source(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	struct bass_recv_state_internal *internal_state = NULL;
+	struct bt_bass_recv_state *state;
+	bt_addr_t *addr;
+	uint8_t pa_sync;
+	uint16_t pa_interval;
+	uint32_t aggregated_bis_syncs = 0;
+
+	/* subtract 1 as the opcode has already been pulled */
+	if (buf->len < sizeof(struct bt_bass_cp_add_src) - 1) {
+		BT_DBG("Invalid length %u", buf->size);
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(bass_inst.recv_states); i++) {
+		if (!bass_inst.recv_states[i].active) {
+			internal_state = &bass_inst.recv_states[i];
+			break;
+		}
+	}
+
+	if (internal_state == NULL) {
+		BT_DBG("Could not add src");
+		return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+	}
+
+	state = &internal_state->state;
+
+	state->src_id = next_src_id();
+	state->addr.type = net_buf_simple_pull_u8(buf);
+	if (state->addr.type > BT_ADDR_LE_RANDOM) {
+		BT_DBG("Invalid address type %u", state->addr.type);
+		return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+
+	}
+
+	addr = net_buf_simple_pull_mem(buf, sizeof(*addr));
+	bt_addr_copy(&state->addr.a, addr);
+
+	state->adv_sid = net_buf_simple_pull_u8(buf);
+	if (state->adv_sid > BT_GAP_SID_MAX) {
+		BT_DBG("Invalid adv SID %u", state->adv_sid);
+		return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+
+	}
+
+	state->broadcast_id = net_buf_simple_pull_le24(buf);
+
+	pa_sync = net_buf_simple_pull_u8(buf);
+	if (pa_sync > BT_BASS_PA_REQ_SYNC) {
+		BT_DBG("Invalid PA sync value %u", pa_sync);
+		return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+	}
+
+	pa_interval = net_buf_simple_pull_le16(buf);
+
+	state->num_subgroups = net_buf_simple_pull_u8(buf);
+	if (state->num_subgroups > CONFIG_BT_BASS_MAX_SUBGROUPS) {
+		BT_WARN("Too many subgroups %u/%u",
+			state->num_subgroups, CONFIG_BT_BASS_MAX_SUBGROUPS);
+		return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+	}
+
+	for (int i = 0; i < state->num_subgroups; i++) {
+		struct bt_bass_subgroup *subgroup = &state->subgroups[i];
+		uint8_t *metadata;
+
+		if (buf->len < (sizeof(subgroup->bis_sync) + sizeof(subgroup->metadata_len))) {
+			BT_DBG("Invalid length %u", buf->size);
+			return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+		}
+
+		subgroup->requested_bis_sync = net_buf_simple_pull_le32(buf);
+
+		if (subgroup->requested_bis_sync &&
+		    pa_sync == BT_BASS_PA_REQ_NO_SYNC) {
+			BT_DBG("Cannot sync to BIS without PA");
+			return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+		}
+
+		/* Verify that the request BIS sync indexes are unique or no preference */
+		if (subgroup->requested_bis_sync & aggregated_bis_syncs &&
```

Comment:
move to function, as for `valid_bis_sync`?

Reply:
Added `bis_syncs_unique_or_no_pref`